File: 프로그래머스/lv3/43164. 여행경로/여행경로.py
import copy
result = []
def dfs(idx, ticket, visit, tickets, airport):
    if len(airport) == len(tickets):
        airport.append(ticket[1])
        result.append(copy.deepcopy(airport))
        return
    
    for i, city in enumerate(tickets):
        if city[0] == ticket[1] and not visit[i]:
            visit[idx] = True     
            airport.append(city[0])
            dfs(i, city, visit, tickets, airport)
            airport.pop()
            visit[idx] = False
    
    return 
def solution(tickets):
    tickets.sort(key = lambda x:(x[0], x[1]))
    for idx, ticket in enumerate(tickets):
        if ticket[0] == "ICN":
            airport = ["ICN"]
            visit = [False] * len(tickets)
            dfs(idx, ticket, visit, tickets, airport)
    return result[0]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;
class Solution {
    static List<Integer> idx = new ArrayList<>();
    static List<String> strings = new ArrayList<>();
    static List<String[]> answers = new ArrayList<>();
    static HashMap<String, Integer> map = new HashMap<>();
    static boolean[] visited;
    static int[] order; 
    static int len;

    static void dfs(int cnt, int prev, String[][] tickets){
        if(cnt == len){
            String[] tmp = new String[len + 1]; 
            StringBuilder sb = new StringBuilder();
            for(int i = 0; i < len; i++){
                tmp[i] = tickets[order[i]][0];
                sb.append(tmp[i]);
                if(i == len - 1)
                    tmp[i + 1] = tickets[order[i]][1];
            }
            
            map.put(sb.toString(), answers.size());
            strings.add(sb.toString());
            answers.add(tmp);
            return;
        }
        for(int i = 0; i < len; i++){
            if(!visited[i] && tickets[prev][1].equals(tickets[i][0])){
                visited[i] = true;
                order[cnt] = i;
                dfs(cnt + 1, i, tickets);
                visited[i] = false;
            }
        }
    }
    
    public String[] solution(String[][] tickets) {
        len = tickets.length;
        for(int i = 0; i < len; i++){
            if(tickets[i][0].equals("ICN")){
                idx.add(i);
            }
        }
        
        order = new int[len];
        visited = new boolean[len];
        for(int start: idx){
            visited[start] = true;
            order[0] = start;
            dfs(1, start, tickets);
            visited[start] = false;
        }
        Collections.sort(strings);
        return answers.get(map.get(strings.get(0)));
    }
}